api/src/main/java/jakarta/mail/util/SharedFileInputStream.java

```diff
@@ -84,6 +84,12 @@ public class SharedFileInputStream extends BufferedInputStream
      */
     private boolean master = true;
 
+    /**
+     * Reference to the top level stream in order not to be junked by garbage collection when there is a remaining derived stream.
+     * null if this is a top level stream
+     */
+    private SharedFileInputStream masterInputStream = null;
```

Comment:
If this.close() is invoked I would think that the reference to top level stream should be set to null but not closed.

Reply:
You are talking about when the derived stream is closed, Right?

Totally agree with that.

I understand you suggest I set `this.masterInputStream` as `null` when `this.close()` of derived stream is invoked but do not let it close `RandomAccessFile`.

But I think when `this.close()` of the derived stream is invoked, the `RandomAccessFile` is not closed because `SharedFile.close()` is already handling it.

If the stream is not master, `this.close()` invokes `SharedFile.close()`. 

`SharedFile.close()` is like below

```
public synchronized void close() throws IOException {
	    if (cnt > 0 && --cnt <= 0)
		in.close();
	}
```

As you can see, It does not close `RandomAccessFile` except when `cnt` is 1, which means it is the last object to have an access to the `SharedFile` or `RandomAccessFile`.

So, I guess I only need to add a code to set `this.masterInputStream`as `null` when `this.close()` of derived stream is invoked.

Anything wrong I misunderstood?

Reply:
Correct.  Essentially we can just unconditionally null this.masterInputStream as the last statement:

````
 @Override
    public void close() throws IOException {
        if (in == null)
            return;
        try {
            if (master)
                sf.forceClose();
            else
                sf.close();
        } finally {
            sf = null;
            in = null;
            buf = null;
            masterInputStream = null;
        }
    }
````

I'm thinking that if we are pinning the root stream by holding a strong reference on creation of a child stream then when the child is closed the link from child SharedFileInputStream to the root SharedFileInputStream should be unlinked (set to null).  Not that we are leaking a file handle but just that we are retaining a small amount of garbage if the child stream is held on to by the caller (3rd party code).

Reply:
Got it. I misunderstood you were saying my code causes the derived stream to close the `RandonAccessFile`   😂

I will edit `this.close()` like you suggested. It is exactly the same way I was about to do 😎